### PR TITLE
OLED integration

### PR DESCRIPTION
**Pull request by moinmoin-sh**
I think that should work. Polished the added files. Made a small change to get_ and set_invert to support the uint16_t.
I hope that I didn't break anything in your logic - should work when no OLED found. Any update on your plans to get a working OLED? Sorry for the poor quality of the pictures.
IMG_E0794
IMG_E0792

### CODE REVIEW

File: Firmware/USMArduinoFirmware/config.h.example [hunk 1]
```diff
@@ -40,4 +40,15 @@ const uint8_t mqtt_lwt_retain       = 1;
...
 
+/* OLED - ONE and ONLY ONE OLED_TYPE should be defined (comment out an others) */
+//#define       OLED_TYPE_SSD1306                           // mostly the .96"
+#define       OLED_TYPE_SH1106                              // mostly the 1.3"
+
+#define       OLED_RESET              -1                    // Reset pin (or -1 if sharing Arduino reset pin)
+#define       OLED_CONTRAST_DIM     5                       // OLED contrast when dimmed
+#define       OLED_CONTRAST_ON      128                     // OLED contrast when normal (on)
+#define       OLED_TIME_ON          10000                   // OLED time in msec with brightness ON after trigger
+#define       MAX_SHOW_EVENT_TIME   3000                    // the time in ms the event line is shown if no new event occured
```
Comment (sumnerboy12) [inline Firmware/USMArduinoFirmware/config.h.example:51]: Prefix with `OLED_` for consistency with other variables?

File: Firmware/USMArduinoFirmware/USMArduinoFirmware.ino [hunk 2]
```diff
@@ -617,7 +702,33 @@ void scanI2CBus()
       }
       else
       {
-        Serial.println(F("ignored"));
+//        Serial.println(F("ignored"));
```
Comment (sumnerboy12) [inline Firmware/USMArduinoFirmware/USMArduinoFirmware.ino:705]: Remove commented code

File: Firmware/USMArduinoFirmware/USMArduinoFirmware.ino [hunk 3]
```diff
@@ -617,7 +702,33 @@ void scanI2CBus()
...
+        g_oled_found = 1;
```
Comment (sumnerboy12) [inline Firmware/USMArduinoFirmware/USMArduinoFirmware.ino:706]: Have to be careful here, since `MCP_MAX_COUNT` is configurable. You could end up in here if someone set that variable to 4. 

We probably need to explicitly check for the OLED at the specific address, outside of this loop, and leave this loop for MCP scanning only.

**Review by sumnerboy12 — COMMENTED** [review]
I haven't tested it yet (bit busy here at the moment), but if you get a chance to improve the I2C scanning I will test once you are happy with it.

Thanks for doing this! The screenshots look great.

File: Firmware/USMArduinoFirmware/USMArduinoFirmware.ino (hunk 3 above)
Comment (moinmoin-sh) [reply Firmware/USMArduinoFirmware/USMArduinoFirmware.ino:706]: I had the same concerns with the I2C scan but dicided to stay with what you had introduced.
Will seperate the OLED check as you suggested and fix the other issues later today.
Thanks

Comment (moinmoin-sh) [reply Firmware/USMArduinoFirmware/USMArduinoFirmware.ino]: I hope that I got i right and you see the updates that I made

Comment (sumnerboy12) [reply Firmware/USMArduinoFirmware/USMArduinoFirmware.ino]: I just compiled it for my Uno and get a low memory warning if `MCP_MAX_COUNT` is 8. Dropping it to 6 compiles with no warning. But I think that is ok for now, I will merge this - thanks very much!

File: Firmware/USMArduinoFirmware/config.h.example [hunk 4]
```diff
@@ -35,9 +35,21 @@ const uint8_t mqtt_lwt_retain       = 1;
 #define       MCP_MAX_COUNT           8                     // Max number of supported MCP23017 chips 
 
 /* Watchdog */
-#define       ENABLE_WATCHDOG         false
```
Comment (sumnerboy12) [inline Firmware/USMArduinoFirmware/config.h.example:38]: Can you revert this change? All the other on/off vars are `ENABLE_XXX` so I would like to keep them consistent (sorry for my OCD!).

File: Firmware/USMArduinoFirmware/config.h.example [hunk 5]
```diff
@@ -35,9 +35,21 @@ const uint8_t mqtt_lwt_retain       = 1;
...
+#define       WATCHDOG_ENABLE         false
 #define       WATCHDOG_PIN            3                     // Output to pat the watchdog
 #define       WATCHDOG_PULSE_MS       50
 #define       WATCHDOG_RESET_MS       30000
 
+/* OLED - ONE and ONLY ONE OLED_TYPE should be defined (comment out an others) */
+//#define       OLED_TYPE_SSD1306                           // mostly the .96"
```
Comment (sumnerboy12) [inline Firmware/USMArduinoFirmware/config.h.example:44]: I wonder if we should have an `ENABLE_OLED` here to keep the config consistent with other parts?

File: Firmware/USMArduinoFirmware/USMArduinoFirmware.ino [hunk 6]
```diff
@@ -86,11 +86,15 @@
...
 #include "Adafruit_MCP23017.h"        // For MCP23017 I/O buffers
 #include "USM_Input.h"                // For input handling (embedded)
+#include "USM_Oled.h"                 // For OLED runtime displays
 
 /*--------------------------- Constants ----------------------------------*/
 // Each MCP23017 has 16 inputs
 #define MCP_PIN_COUNT     16
 
+// OLED
+SSD1306AsciiWire oled;
```
Comment (sumnerboy12) [inline Firmware/USMArduinoFirmware/USMArduinoFirmware.ino:96]: Could you move this down to the other global objects - i,e. `mcp23017` and `usmInput` arrays, `ethernet` and `mqtt_client`?

**Review by sumnerboy12 — COMMENTED** [review]
Just a couple of very minor things. Then I will merge. Thanks!

File: Firmware/USMArduinoFirmware/config.h.example (hunk 5 above)
Comment (moinmoin-sh) [reply Firmware/USMArduinoFirmware/config.h.example:44]: Thanks for reviewing the changes.
I would  just replace the
 "#ifdef OLED_I2C_ADDRESS   ... #endif"  with 
"if (ENABLE_OLED) { ... }" in I2C-scan to make it consistant. For the other calls to oled I would like to stay with g_oled_found.
Does that align with your conception?

The other changes are understood and I will take care of.

Comment (sumnerboy12) [reply Firmware/USMArduinoFirmware/config.h.example:44]: Yeah perfect - just nice to keep all the config options consistent for newcomers to the code.

File: Firmware/USMArduinoFirmware/USMArduinoFirmware.ino (hunk 3 above)
Comment (moinmoin-sh) [reply Firmware/USMArduinoFirmware/USMArduinoFirmware.ino]: That's weired. I have an original UNO in he meantime and working on that. I'm just below the 75% warning with 8 MCP's 

Sketch uses 30100 bytes (93%) of program storage space. Maximum is 32256 bytes.
Global variables use 1530 bytes (74%) of dynamic memory, leaving 518 bytes for local variables. Maximum is 2048 bytes.

May be we have slightly different libraries installed.

**Comment by sumnerboy12** [comment]
Fantastic, thanks Moin!
